core/src/main/scala/org/apache/gearpump/util/Graph.scala

```diff
@@ -355,19 +393,7 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
    * check whether there is a loop
    */
   def hasCycle(): Boolean = {
-    @tailrec
-    def detectCycle(graph: Graph[N, E]): Boolean = {
-      if (graph.edges.isEmpty) {
-        false
-      } else if (graph.vertices.nonEmpty && !graph.vertices.exists(graph.inDegreeOf(_) == 0)) {
-        true
-      } else {
-        graph.removeZeroInDegree
-        detectCycle(graph)
-      }
-    }
-
-    detectCycle(copy)
+    tryTopologicalOrderIterator.isFailure
```

Comment:
can we cache result here ?

Reply:
The graph is mutable so we have to check whether the Graph is changed if caching the result.

Reply:
sure, this can be left as a future optimization and I'm not a big fan of "mutable"